File: spring-webflux/src/main/java/com/example/webflux/persistence/StreamingWorkRepository.java
package com.example.webflux.persistence;

import com.example.webflux.observability.EventLog;
import org.springframework.r2dbc.core.DatabaseClient;
import org.springframework.stereotype.Repository;
import reactor.core.publisher.Mono;

import java.util.concurrent.atomic.AtomicInteger;

public interface StreamingWorkRepository {
    Mono<Void> stream(String req);
}

@Repository
class StreamingWorkRepositoryImpl implements StreamingWorkRepository {

    private final DatabaseClient databaseClient;

    StreamingWorkRepositoryImpl(DatabaseClient databaseClient) {
        this.databaseClient = databaseClient;
    }

    @Override
    public Mono<Void> stream(String req) {
        return Mono.defer(() -> {
            EventLog.log(req, EventLog.Stage.DB, "stream_started");
            AtomicInteger count = new AtomicInteger(0);
            return databaseClient
                    .sql("SELECT n, pg_sleep(0.0001) FROM generate_series(1, 30000) AS n")
                    .fetch()
                    .all()
                    .doOnNext(row -> {
                        int c = count.incrementAndGet();
                        if (c % 1_000 == 0)
                            EventLog.log(req, EventLog.Stage.DB, "rows_consumed", "count=" + c);
                    })
                    .doOnComplete(() -> EventLog.log(req, EventLog.Stage.DB, "stream_completed",
                            "total=" + count.get()))
                    .doOnCancel(() -> EventLog.log(req, EventLog.Stage.DB, "cancellation_detected",
                            "phase=stream rows_so_far=" + count.get()))
                    .then();
        });
    }
}

File: aspnet/src/Persistence/StreamingWorkRepository.cs
using Npgsql;
using static src.Observability.EventLog;

namespace src.Persistence;

public interface IStreamingWorkRepository
{
    Task StreamAsync(string req, CancellationToken ct);
}

public class StreamingWorkRepository(NpgsqlDataSource dataSource) : IStreamingWorkRepository
{
    public async Task StreamAsync(string req, CancellationToken ct)
    {
        NpgsqlConnection conn;
        try
        {
            conn = await dataSource.OpenConnectionAsync(ct);
        }
        catch (Exception) when (ct.IsCancellationRequested)
        {
            Log(req, Stage.Db, "cancellation_detected", "phase=connection");
            throw new OperationCanceledException(ct);
        }

        await using var _ = conn;
        Log(req, Stage.Db, "stream_started");

        try
        {
            await using var cmd = new NpgsqlCommand(
                "SELECT n, pg_sleep(0.0001) FROM generate_series(1, 30000) AS n", conn);
            await using var reader = await cmd.ExecuteReaderAsync(ct);

            int count = 0;
            while (await reader.ReadAsync(ct))
            {
                count++;
                if (count % 1_000 == 0)
                    Log(req, Stage.Db, "rows_consumed", $"count={count}");
            }

            Log(req, Stage.Db, "stream_completed", $"total={count}");
        }
        catch (Exception) when (ct.IsCancellationRequested)
        {
            Log(req, Stage.Db, "cancellation_detected", "phase=stream");
            throw new OperationCanceledException(ct);
        }
    }
}
